### PR TITLE
refactor and reorder checkout processing

### LINKED ISSUE #3363: Checkout API can create a user account when order fails due to coupon restriction (or other failures)
## Describe the bug
~~This affects stores with no payment methods set up, with checkout block configured to allow sign-up on checkout. If the shopper checks `Create an account` and submits the order, the account is created, even though the order isn't processed because of no valid payment method.~~

There may be similar issues, e.g. if a payment fails.

**Update** Refocusing this bug on a specific case for now. When a restricted coupon is used, an account is created and associated with the draft order. 

## To reproduce
See [comment below for test setup & steps](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/3363#issuecomment-728387054).

Previous "No payment methods" steps:

1. Add a checkout block to a published page with Allow shoppers to sign up for a user account during checkout option enabled.
2. In `WooCommerce > Settings > Payments` disable all payment methods.
4. On front end, logged out (or incognito) add something to cart and proceed to checkout.
4. Complete checkout form and select Create an account.
4. Submit.

## Expected behavior
~~No account created unless order succeeds. The order processing and any related logic/entities should only happen if the order succeeds (or preflight checks confirm that the order will succeed.~~

- Account should only be created when associated with a successful order. No account should be created for `checkout-draft` order status.
- Store API should behave predictably and consistently (TBD!) when creating accounts and progressing order state. There may be some differences with legacy / shortcode checkout processing.

### PR DESCRIPTION
**Pull request by haszari**
Fixes #3363

Related: #3260 #3423  https://github.com/woocommerce/woocommerce-subscriptions/pull/3825

This PR revamps `POST /checkout` processing in the Store API to address various issues and inconsistencies. 

The overall order of processing is as follows (more detail tbc):

1. Obtain Draft Order
1. Process Request
1. Process Customer
1. Validate Order
1. Process Payment

The following expectations apply to order status:

- Order should be `checkout-draft` status until payment is attempted.
- Order should only be `pending` once initial validation and processing has been attempted; it should be set to `pending` when it is _ready_ for payment processing.

The following expectations apply to checkout signup (aka create account):

- Customer accounts can be created as a side-effect of order processing; a checkout request can fail, yet still sign up the shopper and log them in.
- If an account is created:
  - Customer should be logged in immediately.
  - Customer id should be returned with response (even if it's an error) so client can update user-related UI. 
  - Any open / existing draft orders for the session should be updated with customer id.
  - Store API nonce should be refreshed and returned with response.
- If account creation fails:
  - Order should not proceed to payment; error should be returned, order should remain at `checkout-draft`. 

Note - not all the above account create items are implemented, though there is related work in #3429. (This might be handled as follow up once the key checkout processing is fully tested.)

##### Order date should match when order was "placed"

The order date is when draft order was created. If order sits around for a while, this could be misleading (e.g. up to a day out).

I'm wondering if we should reset the date when the order goes `pending`. We can probably log an issue and handle as follow up (not a blocker to this PR IMO). 

##### Returning customer id and updated nonce with error responses
We need to return the customer ID (and a new nonce) when customer account is created, in particular when there's an error (e.g. #3429). This PR implements this, and handling of the nonce client side. 

This is a basic fix for #3429 – this flow now works, but there's more UX polish to add (inform user that they have an account, remove `Create account` checkbox, etc). We can handle this as follow up in #3429.

##### Require an account to checkout
This is handled as before in `CreateAccount` with `$checkout->is_registration_required()` which checks well-known core options and filters.

### How to test the changes in this Pull Request:
Test all purchase / checkout flows, with and without user account, with various settings and errors, with all payment methods (!).

Please consider the following variations and test a good range of combinations:

- Order fail due to bad coupon.
- Order fail due to no stock.
- Order fail due to server-side payment error. (May be simplest to hack this in PHP, e.g. force a fail in BACS/COD.)
- Guest/anon orders or logged in.
- Various account settings: allow guest | require account, allow signup (x shopper `Create an account`=yes/no).

Since this is a revamp of checkout processing, I'd appreciate some extra care & time testing! Reviewers please leave a summary of the use-cases you tested. And please add ideas for variations that I've missed (and test 'em!).

### Changelog

> Fix: Fixed a bug in Checkout block (Store API) causing checkout to fail when using an invalid coupon and creating an account.
> Dev: Refactored and reordered Store API checkout processing to handle various edge cases and better support future extensibility.

### CODE REVIEW

**Comment by senadir** [comment]
> The new hook for Woo Subs use case (__experimental_woocommerce_blocks_checkout_order_processed) is still present for now, but will not be needed if Woo Subs switches to using standard Woo core filters (PR in progress - woocommerce/woocommerce-subscriptions#3883).

**This hook has nothing to do with account creation**, it is used for creating subscriptions from a cart that contains subscription products.

What woocommerce/woocommerce-subscriptions#3883 solves is mainly moving from using a custom complicated account forcing mechanism to just switching the value of `checkout->is_registration_required()` and letting Checkout handle how it can be forced.

**So we shouldn't delete that action if the above PR is merged.**

**Comment by mikejolley** [comment]
r/e the `customer_id` response on fail, we _could_ do this for the `checkout` route like what is done in the abstract/cart — override `get_route_error_response` and return it there. Is that what you tried?

The other thing would just be to ensure `X-WC-Store-API-Nonce` is returned as a header when receiving an error so our middleware can catch it.

**Comment by haszari** [comment]
> So we shouldn't delete that action if the above PR is merged.

Woops - thanks for pointing that out @senadir , I was confusing these new hooks! I'll update the comment to clarify the purpose, and keep this in mind when testing.

Do you think we should move this down into step 4, after validating the order (coupons etc), immediately before payment?

**Comment by haszari** [comment]
> r/e the customer_id response on fail, we could do this for the checkout route like what is done in the abstract/cart — override get_route_error_response and return it there. Is that what you tried?

I didn't try that exactly @mikejolley - I did it in `AbstractRoute`, for all routes (!), using global get logged in user func to get the id.

I'll revisit this today, override `get_route_error_response` in checkout, track the newly created customer ID as a member, and return it when there's a new customer.

> The other thing would just be to ensure `X-WC-Store-API-Nonce` is returned as a header when receiving an error so our middleware can catch it.

Agreed, I'll use the approach in #3429 - refresh the nonce near the end of request processing.

Sounds like it's best to handle returning the ID and valid nonce as part of this PR, and then once this is merged, can rebase #3429 to fix up the logic on client.

**Comment by haszari** [comment]
Updates…

#### Working well
- Now firing `__experimental_woocommerce_blocks_checkout_order_processed`. @senadir I've tweaked the comment for the hook to add more info, and tested buying a subscription - seems to work the same as current `trunk`. 🎉 
- Returning the `user_id` for all checkout error responses, via `Checkout::get_route_error_response()`. This seems simple and safe. Will need JS changes to make use of it (similar to #3429). 🎉 

#### Blocker
- ❌ Having issues refreshing the nonce for error requests. 

The problem here is that the way to set the nonce is to call a method on `WP_Rest_Response` - `$response->header( 'X-WC-Store-API-Nonce' … )`. When we're returning an error, e.g. coupon validation, or no stock, we may still need to return a new nonce (because customer account may have been created => login change). 

When the request is an error, we typically return `WP_Error` and WP converts that into a response for us. I.e. we don't have a response to call `->header()` on. 

@mikejolley  Do you know how we can solve this? 

One idea would be that we do our own error responses manually, instead of returning `WP_Error`, especially for common/expected errors like coupon validation. 

Or there may be another PHP mechanism for setting header that works well for this case. 

Note - this problem affects #3429 as well.

**Comment by mikejolley** [comment]
@haszari I've not tested the full flow, but I've pushed a commit which exposes the nonce on all API responses, including Error responses. I had to borrow some code from WP core to make it happen but that should cater for our needs.

**Comment by senadir** [comment]
Note that I can't add products from All Products for guests (X-WC-Store-API-Nonce is invalid.)

**Comment by senadir** [comment]
I'm also going to copy the same comment I had in #3423, this change would cause regression and will undo #3260.
The order object will have unsaved data (customer ID) so code hooking into __experimental_woocommerce_blocks_checkout_order_processed will not have access to the latest order data.
image

#3260 fixed this issue by changing the order so that we save the Order after we finished creating an account but just before triggering the hook.

You can keep this PR as is (with minimal change) either by:
- Saving the order when you create a customer (therefore saving to DB twice in a row, which is not ideal).
- Not saving the order in `update_order_from_request` and doing one save before triggering the action.

Of course, WooCommerce Subscription and other plugins can simply use the object passed to them instead of reaching for `wc_get_order( $order->get_id() )` but we already discussed why this isn't ideal and would create bugs down the road for other plugins.

**Comment by haszari** [comment]
> I've pushed a commit which exposes the nonce on all API responses, including Error responses. I had to borrow some code from WP core to make it happen but that should cater for our needs.

Awesome, thanks for that @mikejolley ! I've added a fix to catch this in the client (bd49ccca0050a5c6d505d8b75f16e55d03370db5) and now the "create account + invalid coupon" flow works correctly: submit => (account created) bad coupon message, removed from order => resubmit => success 🎉 

So that's that part is functional now. We can add more UX polish in #3429 as follow up. cc @opr who was excited about making this nicer with a "you have an account" notice :)

**Comment by haszari** [comment]
> this change would cause regression

@senadir Just to clarify the issue here, I looked at the PR (no issue) and I think this is the bug - let me know if I've missed anything:

- Checkout block + WooCommerce Subscriptions enabled. Store should have a subscription product set up.
- Clone [this branch from subscriptions](https://github.com/woocommerce/woocommerce-subscriptions/pull/3825). This hooks up subscriptions order processing to create a subscription for blocks/Store API orders.
- As a guest / anon user, add a subscription product to cart.
- Ensure an account will be created - either store settings, or check `Create an account` in front end.
- Complete checkout and submit. Recommend using Stripe CC to pay (I think there are issues with other payment methods).
- Expected: new customer account, successful order, subscription created alongside order.
- Actual: order fails with `Invalid subscription customer_id.`

Screen Shot 2020-11-26 at 9 52 07 AM

Sounds like the simplest fix (at least for now) is to `save()` the order before firing the hook; it seems sensible to ensure that the passed `$order` matches the state in DB before calling any hook. Maybe we can optimise this later if needed, or require that extensions only use the order passed to the hook (I'm not sure of the tradeoffs with that approach).

I added this extra save (9933c2008603beda56e06a84b91036805e8eb365) and tested (with subs branch) and confirmed that use case (steps above) works correctly. 

> Note that I can't add products from All Products for guests (X-WC-Store-API-Nonce is invalid.)

@senadir Are you referring to [`All Products for WooCommerce Subscriptions`](https://woocommerce.com/products/all-products-for-woocommerce-subscriptions/?quid=46bb4b499eb4859ee19ae47471d5364b) or something else? Perhaps try again with the new nonce fix - it was sent, but not refreshed in the client-side store (bd49ccc).

If there is still an issue, perhaps we can work on this as follow up, since Subscriptions / extensions support is a work in progress right now.

**Comment by haszari** [comment]
This is ready for review now! Would appreciate a thorough review (or 2) and lots of edge-case testing, since this is a moderate change to a critical API (checkout processing).

**Comment by senadir** [comment]
> Are you referring to All Products for WooCommerce Subscriptions or something else? Perhaps try again with the new nonce fix - it was sent, but not refreshed in the client-side store (bd49ccc).

No, I was referring to our All Products block.
It works fine now.

**Comment by senadir** [comment]
I tested the current implementation and it works fine (with subs).
So I went ahead and removed `$order->save()` from `update_order_from_request` so that we only save once, this still worked fine and produced no issues because we never reach for the DB in that function.

Now, the issue we're solving by introducing another `save` call could be introduced again if we decide to create another action in that flow, but I don't any reason to introduce a new action there and prefer to solve this issue once we decide to add a new action.

So I'd like to proceed with the above suggestion (deleting [this line]( https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/9933c2008603beda56e06a84b91036805e8eb365/src/StoreApi/Routes/Checkout.php#L467)) and do some testing on it as well.

**Comment by haszari** [comment]
> removed $order->save() from `update_order_from_request` so that we only save once, this still worked fine and produced no issues because we never reach for the DB in that function.
> So I'd like to proceed with the above suggestion (deleting this line) and do some testing on it as well.

This makes sense - I can commit that change today and do some testing to be sure. It might be more robust to leave it in place, I'm not clear on the performance implications (keen for review feedback on that). @senadir If I don't get to this before I go AFK, please feel free to commit to this PR :)

File: assets/js/base/context/cart-checkout/checkout/processor/index.js
Comment (nerrad) [inline assets/js/base/context/cart-checkout/checkout/processor/index.js]: I'm curious why the change on argument name here? `error` was used because this is a catch callback (so it's just a nudge that this is handling an error response). If anything, should the variable name be changed I would suggest `errorResponse` instead.

**Comment by haszari** [comment]
@senadir Regarding removing that extra `$order->save()`, I'm less confident that we should attempt to optimise now.

Tracing through the logic in `Checkout::get_route_post_response()` (and subroutines) I see multiple places where we modify `$order` or `$customer` object, and a multiple places where we (re)save them too. 

For example, `$customer` is saved in both [`Checkout::update_customer_from_request()`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/fix/checkout-signup-order-processing-logic/src/StoreApi/Routes/Checkout.php#L449) (refactored on this branch) and  [`OrderController::sync_customer_data_with_order()`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/fix/checkout-signup-order-processing-logic/src/StoreApi/Utilities/OrderController.php#L96).

I agree that it would be better to coordinate saving these entities to the database only once per POST request, but I'm not sure if that's practical, since we will likely have more extensibility hooks over time (and extensions might assume everything is persisted to the database when a hook is fired).

Or put another way - I don't think it should block this PR; we could follow up to ensure we are using a consistent approach, and ensure that approach isn't causing a performance issue (or an extensibility issue!). Interested to hear feedback from reviewers about this (cc @mikejolley).

**Update**: Just realised while working on #3429 - checkout processing can be aborted at any stage if anything throws an exception. So it's probably a good idea for us to persist objects early and often, so any changes are saved even if there is an error.

**Comment by opr** [comment]
I've done some testing with various conditions, here are the results:

- ✅ Coupon on order with ineligible email - account created and order does not succeed
- ❌ Cannot proceed after this still because of the Invalid cookie nonce error
- ✅ Checkout without choosing to create account - succeeds and no account is created
- ✅ Checkout while opting to create account - order succeeds and account is created
- ✅ Stripe card declined - account created and order not created
- ⚠️ Out of stock product in order, good because the order doesn't succeed but small issue with how the error displays (&amp;quot; appears around the product name) Screenshot 2020-11-26 at 21 19 51
- ^ #3469 will take care of this

**Comment by haszari** [comment]
Thanks for testing!

> Cannot proceed after this still because of the Invalid cookie nonce error

@opr Can you list exact steps for reproducing this? I've tested and confirmed this is fixed with latest changes, maybe I missed a use case. Make sure you're on the latest code and have rebuilt (this PR has been rebased recently).

**Comment by opr** [comment]
> Can you list exact steps for reproducing this? I've tested and confirmed this is fixed with latest changes, maybe I missed a use case. Make sure you're on the latest code and have rebuilt (this PR has been rebased recently).

Sure, I did run `npm build` too, so I think everything is correct.

1. Create a coupon with an email restriction (I used _*@test.com_)
2. Ensure users can create accounts at checkout at block level and store level.
3. Go to the front-end and add two products, (I chose beanie and beanie with logo from the default products)
4. Go to a page with the checkout block, and apply the coupon from step 1.
4. Fill in all fields, and tick "Create account", ensuring your email is not valid for the coupon
5. Try to checkout, and get an error message.
6. Try to immediately check out again

See a video of the process here: https://www.loom.com/share/fe82275fada14f2cbc99ee34800b55ad

**Comment by haszari** [comment]
Dang - I can't reproduce that @opr - maybe there's something hinky happening with a build. Here's a walkthrough from 21e5cd8703730d76563845ad68bf220f977dfa5f on this branch. 🤞🏻I'm hoping this was a stale JS build in your test environment…

https://d.pr/i/ooWTix (big gif, no embed)

**Comment by nerrad** [comment]
I just tested to try and reproduce the invalid cookie nonce error as well and was unable to reproduce so it appears to be fixed from my testing.

File: src/StoreApi/Routes/AbstractRoute.php
Comment (nerrad) [inline src/StoreApi/Routes/AbstractRoute.php]: Are we sure we want to return a nonce header for _every_ error response?

File: src/StoreApi/Routes/AbstractRoute.php
Comment (nerrad) [inline src/StoreApi/Routes/AbstractRoute.php]: I'm not sure on the need for this abstraction right now. Do we anticipate returning additional response headers here eventually or will there be cases where we only want a certain set of headers added to a response? Will this be called from a variety of different places? 

I would prefer we avoid adding this abstraction until we know it's necessary - not only that, but current implementation of the method is ambiguous - when seeing this method implemented, the first question I would have is, "what response headers are added"?

File: src/StoreApi/Routes/AbstractRoute.php
Comment (nerrad) [inline src/StoreApi/Routes/AbstractRoute.php:81]: `WP_Rest_Server` [already handles adding `wp_errors` to the response](https://github.com/WordPress/wordpress-develop/blob/69c3ab8a118679b3f560b7f332068e180d6d7310/src/wp-includes/rest-api/class-wp-rest-server.php#L439-L441) why do we need to do it here (and also create our own `error_to_response` method)?

Is it because we need to add the nonce header to the response after the conversion (instead of returning a wp_error)? If so, it looks like we could use the [`rest_post_dispatch` filter](https://github.com/WordPress/wordpress-develop/blob/69c3ab8a118679b3f560b7f332068e180d6d7310/src/wp-includes/rest-api/class-wp-rest-server.php#L463) instead to enhance with the nonce header after the wp_error conversion?

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php]: This method is being used to validate, when it explicitly returns a payment method for the given id. Instead, we should maybe modify `process_payment` so it receives an optional `$payment_method_id` as an argument, then you can retrieve the payment method here and feed it into `process_payment`. Then `process_payment` only needs to call `get_request_payment_method_id` if it's not present.

So `Checkout::process_payment` would have this as its signature:

```php
protected function process_payment( \WC_Order $order, \WP_REST_Request $request, string $payment_method_id  = '' ) { 
    /** ... **/
   $context->set_payment_method( $payment_method_id === '' ? $this->get_request_payment_method_id( $request ) : $payment_method_id );
    /** ... **/
}
```

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php]: I think this is a good initial start. I do wonder however if some extensions might need the request data for any additional information that might have come through in the request. That's unknown at this point though and not something we expose as an option client side for the moment.

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php]: Please follow the [WordPress documentation standards for action/filter hooks](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters). I know we haven't been consistent with this but I'd like us to fix that.

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php]: This is where you'd feed in the `$payment_method_id` if you implement my suggested changes above.

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php:554]: I know this pattern wasn't established in this pr. But as a follow-up issue, I'd like to see us inject `OrderController` and `CartController` as dependencies for this class (I would make it a requirement before making the StoreAPI public).

**Review by nerrad — COMMENTED** [review]
Starting to look pretty good! I'll defer to others regarding the order of processing changes made but I did a review of the implementation/syntax and have a few comments I'd like see addressed.

File: src/StoreApi/Routes/AbstractRoute.php
Comment (mikejolley) [reply src/StoreApi/Routes/AbstractRoute.php:81]: `rest_post_dispatch` potentially runs on every REST response from all APIs. Converting it ourselves gives us full, contained, control over error responses in the Store API without affecting non-store API endpoints.

Also pointing out that feedback above states:

> Are we sure we want to return a nonce header for every error response?

We would need a way to identify store API requests with this filter, and then further if we wanted to only affect *some* Store API requests. Just pointing this out because it's a downside to using the filter.

File: src/StoreApi/Routes/AbstractRoute.php
Comment (mikejolley) [reply src/StoreApi/Routes/AbstractRoute.php]: Yes; we should include the response in all Store API responses because:

a) It gives the client consistent headers to capture through a middleware
b) A login can occur outside of the Store API flow, so theoretically the nonce could refresh at any time (we may need a retry mechanism at some point)

File: src/StoreApi/Routes/Checkout.php
Comment (mikejolley) [reply src/StoreApi/Routes/Checkout.php]: Why doesn't the linting catch this?

File: src/StoreApi/Routes/AbstractRoute.php
Comment (mikejolley) [reply src/StoreApi/Routes/AbstractRoute.php]: As an alternative, if the intention is to split up and group logic, but not abstract this, just make it private?

File: src/StoreApi/Routes/Checkout.php
Comment (senadir) [reply src/StoreApi/Routes/Checkout.php]: Yes, this could happen, the original hook passes `$_POST` data.
However, passing $_POST was required because people who render anything in the checkout form and passing `POST`  would provide a way to catch it.
We're not planning on doing that, and since we do everything via REST, any passed back data should via an explicit field and via an explicit mechanism.
Any core data in `$request` that is not present in the order is data we might not want to expose, and exposing `$request` might be too much implementation detail.
So if there's something that we need from `$request` that is not present in order, we can expose it explicitly in the hook.

File: src/StoreApi/Routes/Checkout.php
Comment (senadir) [reply src/StoreApi/Routes/Checkout.php]: TIL!

File: src/StoreApi/Routes/AbstractRoute.php
Comment (nerrad) [reply src/StoreApi/Routes/AbstractRoute.php:81]: If we hook into the filter only when this route is being processed it will end up only getting processed on this route right? An advantage of using the filter is that if we add it to a late priority (eg `999`) we also ensure the header is added at the last possible moment?

Loosely held opinion, but I like to avoid creating new code unnecessarily if possible here.

File: src/StoreApi/Routes/AbstractRoute.php
Comment (nerrad) [reply src/StoreApi/Routes/AbstractRoute.php]: Explanation makes sense. I just want to make sure we're not exposing an easy way for bad actors to grab nonces for malicious use  - I don't think they can because the request is still tied to either a cart session or an authenticated request.

File: src/StoreApi/Routes/AbstractRoute.php
Comment (nerrad) [reply src/StoreApi/Routes/AbstractRoute.php]: Sure, but the naming is still ambiguous, it's only adding the store api nonce header at the moment. Currently this method is only called in one place and it could easily be replaced with just adding the header to the response in that one place. It seems like premature abstraction for something we might not even use to add additional headers to the response for.

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [reply src/StoreApi/Routes/Checkout.php]: Right, I'm not suggesting we change this signature now. Just thought it was important to note that there may be need for passing along additional information from the request that might not be exposed on the order. It'd be good to think of how we could do that (but not for this PR).

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [reply src/StoreApi/Routes/Checkout.php]: > Why doesn't the linting catch this?

Ya I was surprised too.

File: src/StoreApi/Routes/Checkout.php
Comment (mikejolley) [reply src/StoreApi/Routes/Checkout.php]: I think we can also move validation to the schema (enum property)

File: src/StoreApi/Routes/AbstractRoute.php
Comment (mikejolley) [reply src/StoreApi/Routes/AbstractRoute.php:81]: > An advantage of using the filter is that if we add it to a late priority (eg 999) we also ensure the header is added at the last possible moment?

This is not an advantage against inline header addition which is only done on Store API responses—late addition is of no concern with the current approach.

Another advantage of inline is that our `get_response` methods only ever return valid `WP_REST_Response` objects, rather than __maybe__ returning a `WP_Error` object, so it's more predictable. The code introduced is fairly small and does enable us to further customise the error responses if needed.

File: src/StoreApi/Routes/AbstractRoute.php
Comment (mikejolley) [inline src/StoreApi/Routes/AbstractRoute.php:84]: @haszari this is instead of user_id in responses. It's not consumed anywhere yet right?

Comment (nerrad) [reply src/StoreApi/Routes/AbstractRoute.php:84]: It looks like #3429 would need updated to get the user id.

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php]: We shouldn't need the root `\` here with the import at the top of the file right?

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php]: I know this variable wasn't introduced in this pr but I noticed in reviewing, what purpose does `$created` have in this method? It doesn't seem to be implemented anywhere (other than switched to `true` when creating the order)?

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php]: `Checkout::update_order_from_request` already checks to see if `$request['payment_method']` exists. I think we could move this validation inside there to keep the logic flow in this method cleaner? Also it would help surface this error earlier in the stack because `update_order_from_request` already does some validation around whether a payment method exists for the payment method id passed in the request.

File: src/StoreApi/Routes/Checkout.php
Comment (nerrad) [inline src/StoreApi/Routes/Checkout.php]: I know this was moved here from `handle_error` method so was pre-existing to this pr, however we should probably implement WordPress coding standards for filters here. Might also be worth adding an inline dev note that any html will be escaped on the frontend when this is added to a notice in the blocks (because our notices system escapes html right?).

**Review by nerrad — APPROVED** [review]
Nice bit of refactoring here Mike in simplifying the logic flow, and doing a bit of code cleanup around scope and removing unnecessary methods. 

I have a few comments that are not blocking - you can address as you see fit so I'm pre-approving. However _I did not test_ as it requires more time than I can give at the moment for testing. So up to you whether you are happy with your testing of this pr or if you want to see if someone else can confirm testing.

File: src/StoreApi/Routes/Checkout.php
Comment (mikejolley) [reply src/StoreApi/Routes/Checkout.php]: I think we should remove this filter (I'm going to)—I can see it was added to mirror core, however, the filtered strings are different and the core filter allows for HTML whereas this would be more limited. Since this is an error from the API it does not need to be filterable. The client can change the wording based on the error code cc @haszari

File: src/StoreApi/Routes/Checkout.php
Comment (mikejolley) [reply src/StoreApi/Routes/Checkout.php]: That will need tweaking because it's required on POST but not on UPDATE right now. Ill see if that difference is still needed.

**Pull request merged**
